src/workerd/io/trace.c++

```diff
@@ -5,15 +5,159 @@
 #include <workerd/io/trace.h>
 #include <workerd/util/thread-scopes.h>
 
+#include <openssl/rand.h>
+
 #include <capnp/message.h>
 #include <capnp/schema.h>
+#include <kj/compat/http.h>
 #include <kj/debug.h>
 #include <kj/time.h>
 
 #include <cstdlib>
 
 namespace workerd {
 
+namespace tracing {
+namespace {
+kj::Maybe<kj::uint> tryFromHexDigit(char c) {
+  if ('0' <= c && c <= '9') {
+    return c - '0';
+  } else if ('a' <= c && c <= 'f') {
+    return c - ('a' - 10);
+  } else if ('A' <= c && c <= 'F') {
+    return c - ('A' - 10);
+  } else {
+    return kj::none;
+  }
+}
+
+kj::Maybe<uint64_t> hexToUint64(kj::ArrayPtr<const char> s) {
+  KJ_ASSERT(s.size() <= 16);
+  uint64_t value = 0;
+  for (auto ch: s) {
+    KJ_IF_SOME(d, tryFromHexDigit(ch)) {
+      value = (value << 4) + d;
+    } else {
+      return kj::none;
+    }
+  }
+  return value;
+}
+
+void addHex(kj::Vector<char>& out, uint64_t v) {
+  const char HEX_DIGITS[] = "0123456789abcdef";
+  for (int i = 0; i < 16; ++i) {
+    out.add(HEX_DIGITS[v >> (64 - 4)]);
+    v = v << 4;
+  }
+};
+
+void addBigEndianBytes(kj::Vector<byte>& out, uint64_t v) {
+  for (int i = 0; i < 8; ++i) {
+    out.add(v >> (64 - 8));
```

Comment:
Nit: Before for loop:

out.reserve(out.size() + 8);

Reply:
The vector passed in already has reserved size. Given that I'm moving this existing implementation out from the internal code base I'm going to leave it as is for now. We can take another pass on speeding it up later.